lib/puppet/application/resource.rb

```diff
@@ -236,8 +240,11 @@ def find_or_save_resources(type, name, params)
           resource = Puppet::Resource.new(type, name, :parameters => params)
 
           # save returns [resource that was saved, transaction log from applying the resource]
-          save_result = Puppet::Resource.indirection.save(resource, key)
-          [save_result.first]
+          save_result, report = Puppet::Resource.indirection.save(resource, key)
+          status = report.resource_statuses[resource.ref]
+          raise "Failed to manage resource #{resource.ref}" if status&.failed? && !options[:fail].nil? && options[:fail]
```

Comment:
This check `!options[:fail].nil?` isn't needed, because of the way application options are parsed in https://github.com/puppetlabs/puppet/blob/1a53bf7d66275ad4b4928484bb1314dc32e84677/lib/puppet/application.rb#L186-L188

For example, this is how the resource application checks boolean options for [`--to_yaml`](https://github.com/puppetlabs/puppet/blob/1a53bf7d66275ad4b4928484bb1314dc32e84677/lib/puppet/application/resource.rb#L152) and [`--edit`](https://github.com/puppetlabs/puppet/blob/1a53bf7d66275ad4b4928484bb1314dc32e84677/lib/puppet/application/resource.rb#L163)